packages/core/src/utils/credential.ts

```diff
@@ -35,3 +37,143 @@ export const getCredentialIdentifiers = (credential: CredentialExchangeRecord) =
     schemaId: credential.metadata.get(AnonCredsCredentialMetadataKey)?.schemaId,
   }
 }
+
+/**
+ * Ensures credential has all required metadata cached. If any metadata is missing,
+ * it will be resolved and added to the credential.
+ * 
+ * @param credential - The credential record to ensure metadata for
+ * @param agent - The agent instance for resolving schema/credDef
+ * @param offerData - Optional offer data containing schema_id and cred_def_id
+ * @returns Promise<boolean> - Returns true if metadata was updated, false otherwise
+ */
+export async function ensureCredentialMetadata(
+  credential: CredentialExchangeRecord,
+  agent: Agent,
+  offerData?: { schema_id?: string; cred_def_id?: string }
+): Promise<boolean> {
+  if (!agent?.modules?.anoncreds) {
+    return false
+  }
+
+  const existingMetadata = credential.metadata.get(AnonCredsCredentialMetadataKey)
+  
+  // Determine schemaId and credDefId
+  let schemaId = offerData?.schema_id || existingMetadata?.schemaId
+  let credDefId = offerData?.cred_def_id || existingMetadata?.credentialDefinitionId
+
+  // If we still don't have IDs, try to get them from format data
+  if (!schemaId || !credDefId) {
+    try {
+      const { offer } = await agent.credentials.getFormatData(credential.id)
+      const formatOfferData = offer?.anoncreds ?? offer?.indy
+      if (formatOfferData && typeof formatOfferData === 'object') {
+        schemaId = schemaId || ('schema_id' in formatOfferData ? formatOfferData.schema_id : undefined)
+        credDefId = credDefId || ('cred_def_id' in formatOfferData ? formatOfferData.cred_def_id : undefined)
```

Comment:
Would this be simpler?

```suggestion
        schemaId = schemaId || formatOfferData?.schema_id
        credDefId = credDefId || formatOfferData?.cred_def_id
```

Reply:
had to do slightly different:
```
schemaId = schemaId || (formatOfferData as any).schema_id
credDefId = credDefId || (formatOfferData as any).cred_def_id
```

Reply:
`as any` should be avoided as much as possible

if it's possible for formatData to be undefined or null or a non-object then we should still use optional chaining. If you still have to cast with `as` try to use the correct type, not `any`

Reply:
I imported the AnonCredsCredentialOffer type